Fix bug introduced on drag and drop external files

pjft opened the pull request:
Drag and drop of external (OS filesystem) to subdirectories in the browser would fail on specific cases, mainly when the subdirectory was no longer off the root folder.
This seemed to have been an issue introduced with the subdirectory free space calculation [here](https://github.com/nextcloud/server/commit/f9536b08096ed1c80391af36d33a18198be1fced) and it seems to fail for any subdirectory that doesn't belong to the root folder.

Bug reports:
- https://help.nextcloud.com/t/drag-drop-into-subfolders/120731
- https://github.com/nextcloud/server/issues/24720

I couldn't find any reference on scenarios or quota management that would suggest when a subdirectory's free space would be different to the parent's free space, other than when on the root folder, where subdirectories can be external mounts.

As such, if my understanding is correct (please let me know if it isn't), this calculation can - and should - be made by getting the free space from the first subdirectory in the total path, which caters for all subdirectory scenarios.

Please advise, happy to help improve this.

pjft committed:
Fix bug introduced on drag and drop external files

Drag and drop of external (OS filesystem) to subdirectories in the browser would fail on specific cases, mainly when the subdirectory was no longer off the root folder.
This seemed to have been an issue introduced with the subdirectory free space calculation [here](https://github.com/nextcloud/server/commit/f9536b08096ed1c80391af36d33a18198be1fced) and it seems to fail for any subdirectory that doesn't belong to the root folder.

Bug reports:
- https://help.nextcloud.com/t/drag-drop-into-subfolders/120731
- https://github.com/nextcloud/server/issues/24720

I couldn't find any reference on scenarios or quota management that would suggest when a subdirectory's free space would be different to the parent's free space, other than when on the root folder, where subdirectories can be external mounts.

As such, if my understanding is correct (please review), this calculation can - and should - be made by getting the free space from the first subdirectory in the total path, which caters for all subdirectory scenarios.

Please advise, happy to help improve this.

Co-authored-by: John Molakvoæ <skjnldsv@users.noreply.github.com>
Signed-off-by: pjft <pjft@users.noreply.github.com>

pjft commented:
Apologies, had made a mistake in my assessment. Actually the issue was that the subdirectory had "/" at the beginning, and we were unable to create a model for it. Getting the last subdirectory in the path without the slash works.
I updated the file.

skjnldsv commented on apps/files/js/file-upload.js:
```diff
@@ -997,7 +997,7 @@ OC.Uploader.prototype = _.extend({
 					} else if (upload.getTargetFolder().indexOf(self.fileList.getCurrentDirectory()) === 0) {
 						// Check subdirectory free space if file is uploaded there
-						var targetSubdir = upload._targetFolder.replace(self.fileList.getCurrentDirectory(), '')
+						var targetSubdir = upload._targetFolder.split("/")[upload._targetFolder.split("/").length-1]
```
I don't like this chain, I am afraid of any breaking point in between.
Can we have a failsafe here and throwing/ displaying proper error to the user? :)

pjft replied:
Of course. To be clear, you're suggesting a try/catch that, if anything fails, at least show an error to the user rather than the current ambiguous state, or something different? :)

Thank you for your time!

skjnldsv replied:
Or directly use 
```js
upload._targetFolder.split('/').pop()
```

pjft replied:
Sounds good. From checking the code, _targetFolder is set prior to this operation, and isn't used anywhere else, so we can use pop() as even though it changes the array, the array itself isn't used anywhere else.

pjft replied:
Apologies. It was 8AM, and first thing I was looking at. We create the array at that moment. Nevermind!

pjft commented:
@skjnldsv Thanks for the help - it works great, but from testing a new case now I find that the upload-to-subfolder code path doesn't quite work as intended. I'm happy to try and fix it, but would appreciate guidance.

Specifically, the current issue is that the upload code will check for conflicts against files _in the currently open directory_, so when uploading to a subfolder it risks not detecting the right conflicts if there are any - and it will also trigger a wrong conflict dialog if there are files with the same name in the current folder.

The current code does `self.checkExistingFiles(selection, callbacks);`. Ideally I'd be able to either call this with a comparable object from the subfolder, or I'd have to change the code in checkExistingFiles to compare against the correct fileList, if it's a subfolder.

I am a bit at a loss as to how to go about it, though, but happy to look into it.

pjft commented:
Updated checkExistingFiles with the same conditions to not execute conflict check at that stage, as it'd compare with the incorrect file set. Should be good now.

szaimen reviewed (commented):
Unfortunately jsunit is failing now, do you mind fixing that? :)

pjft commented:
Definitely. It seems that my attempt at correcting the conflict check with the right folder is making a test fail. I thought that my commit earlier today had fixed it, but alas it didn't. I'm looking into it.

Apologies for the hassle.

pjft commented:
So, I'd like some validation here as the way I fixed jsunit was to update the test - which is always awkward.
What I did was:
- The test was creating an upload with a blank targetDir. Technically, this is a different folder than root ("/") - which is effectively where the test is uploading things to - and that's why the test was now failing: we are no longer checking for conflicts with different folders than the one we are uploading to.
- I changed the test to set the targetDir to "/". My assumption for this change is that there is no user-driven interaction that should result in targetDir being blank. Please let me know if that is incorrect. Also, if there is, uploading to (blank path) should technically not be the same as uploading to ("/"). If it is, I suppose I can change the code to treat those the same, but it becomes less... elegant.
- I ran the test without my current changes, and it ran successfully (meaning that the original code passes the new test). I then made my change to avoid the wrong existingFilesCheck comparison, and the runs successfully.

Since I am not fully familiar with this part of the code, I'd appreciate it if someone more knowledgeable than I am to confirm whether this is an adequate change for the test, or if I unknowingly now changed the test in a way that it no longer suits its intended purpose.

Thank you.

PVince81 commented on apps/files/js/file-upload.js:
```diff
@@ -776,6 +776,10 @@ OC.Uploader.prototype = _.extend({
 				// no list to check against
 				return true;
 			}
+			if (upload.getTargetFolder() != fileList.getCurrentDirectory()) {
```
please use strict comparison: `!==`

pjft replied:
Thanks, updated.

PVince81 commented on apps/files/js/file-upload.js:
```diff
@@ -997,7 +1001,8 @@ OC.Uploader.prototype = _.extend({
 						// Check subdirectory free space if file is uploaded there
-						var targetSubdir = upload._targetFolder.replace(self.fileList.getCurrentDirectory(), '')
+						// Retrieve the folder destination name
+						var targetSubdir = upload._targetFolder.split('/').pop()
```
this seems to only take the last component of the folder.

what if targetFolder has more components like "a/b/c/d" ?

pjft replied:
Correct. That seemed to have been the issue prior to it. 

In the following line, `upload.uploader.fileList` represents the fileList in the current folder. As such, via drag and drop you can only be dragging to either <current folder> or a subfolder of the current folder. 

Assuming you are in folder /a/b/c
and you're dragging a file straight into 'd'

`upload.uploader.fileList` will have all the files in /a/b/c .

Calling `upload.uploader.fileList.getModelForFile('a/b/c/d')` will return null, so 
`upload.uploader.fileList.getModelForFile(targetSubdir).get('quotaAvailableBytes')` will raise an exception.
Calling `upload.uploader.fileList.getModelForFile('d')` will return the model for the folder a/b/c/d, and 
`upload.uploader.fileList.getModelForFile(targetSubdir).get('quotaAvailableBytes')` will return the quota for the subfolder we're dragging the file into.

Was that the case you were bringing up? If not, let me know - as I stated, you know more about this code than I do, I am just trying to help here.

Thank you for your time.

pjft replied:
Correct. That seemed to have been the issue prior to it. 

In the following line, `upload.uploader.fileList` represents the fileList in the current folder. As such, via drag and drop you can only be dragging to either <current folder> or a subfolder of the current folder. 

Assuming you are in folder /a/b/c
and you're dragging a file straight into 'd'

`upload.uploader.fileList` will have all the files in /a/b/c .

Calling `upload.uploader.fileList.getModelForFile('a/b/c/d')` will return null, so 
`upload.uploader.fileList.getModelForFile(targetSubdir).get('quotaAvailableBytes')` will raise an exception.
Calling `upload.uploader.fileList.getModelForFile('d')` will return the model for the folder a/b/c/d, and 
`upload.uploader.fileList.getModelForFile(targetSubdir).get('quotaAvailableBytes')` will return the quota for the subfolder we're dragging the file into.

Was that the case you were bringing up? If not, let me know - as I stated, you know more about this code than I do, I am just trying to help here.

Thank you for your time.

skjnldsv replied:
Yes, that's the point, we use the subfolder name when it's dropped into a subfolder

pjft replied:
Thanks. This PR makes it indeed use the subfolder rather than the full path, which was causing problems when checking for available quota.

pjft replied:
Or would you be more comfortable with 

```
upload._targetFolder.replace(self.fileList.getCurrentDirectory()+'/', '')
```

?

That also works. I just don't have a scenario where the `targetSubdir` is not an immediate child of the current folder, hence using pop(), but let me know if there's a preference in the code.

artonge replied:
Yes please, it feels less error-prone :)

```suggestion
						var targetSubdir = upload._targetFolder.replace(self.fileList.getCurrentDirectory()+'/', '')
```

pjft replied:
Thanks! Actually, I just tested this but this actually breaks when we're at the root folder.

When in a subfolder, 
`self.fileList.getCurrentDirectory()` is, say, "/parent", and 
targetFolder is "/parent/target", so 
we want targetSubdir to be "target" only. 
`self.fileList.getCurrentDirectory()+'/'` is "/parent/", which matches and replaces the right string.
This works.

But if in the root folder, 
`self.fileList.getCurrentDirectory()` is, "/", and 
targetFolder is "/target", so 
we want targetSubdir to be "target" only. 
`self.fileList.getCurrentDirectory()+'/'` becomes "//", which doesn't find a match, and so everything breaks from then on.

I am happy to come up with an `if` condition there, if you feel strongly that we should have the latter version, but thought I'd raise this before sending over another change. :)

pjft commented:
Hi all,

Apologies for bothering once again.

Is there anything pending on my end on this PR? I'm happy to hear that there hasn't been time to look into it, but I'd like some guidance as it's not sure whether I'm the one who's expecting to do something right now on it - and if so, what.

Hope you're doing well, and keep up the great work!

artonge commented on apps/files/js/file-upload.js:
```diff
@@ -776,6 +776,10 @@ OC.Uploader.prototype = _.extend({
 				// no list to check against
 				return true;
 			}
+			if (upload.getTargetFolder() !== fileList.getCurrentDirectory()) {
+				// not uploading to the current folder
+				return true;
+			}
```
Why is this needed?

pjft replied:
Thanks for checking! This was explicitly added because if we're not uploading to the current folder, we don't want to detect conflicts with the current folder - which I happened to run into when testing.

Example with these files and folders:

/ [folder]
/a.txt [file]
/target/ [folder]

If you're in `/`, and drag a file "a.txt" to the "target" subfolder, without this check you'd get a conflict warning saying that it'd already exist a file with that name in the destination, which is incorrect.

The actual conflict check will then run for the destination folder. Adding this here prevents that false conflict.

Does that make it clearer, or is there something I'm probably missing - or unwillingly causing with it?

Thanks!

artonge replied:
Thanks for the explanation. So another conflict check will happen for the subfolder?

pjft replied:
Correct, it does trigger one right now - though I am assuming it is server side (which makes sense as the files from the subfolders aren't available on the client-side javascript context).

artonge reviewed (changes requested):
Sorry for the delay here, can you address my comments? :)

pjft commented:
@artonge definitely - thank you for taking the time. Let me know further thoughts - and definitely a preference on that code statement rewrite.

Have a great day! :)

artonge commented:
> Hi @artonge - apologies for bothering. Just received a notification from #29202 around this and came to check. It seems I'm still pending a change requested by you, and apparently another approver.
> 
> I'm not sure you can approve, but if you or anyone else could help me navigate this it'd be very much appreciated. Thank you all, and keep up the great work :)

Thanks for repinging and for the change @pjft ! And sorry for the delay in our review :)

pjft commented:
Not at al - thank you for taking the time. :)

Regards.

szaimen commented:
Backports?

szaimen commented:
/backport to stable22

szaimen commented:
/backport to stable21

pjft commented:
Well, that was easy :D Thank you!
